### PR TITLE
Misc improvements

### PR DESCRIPTION
**Pull request by mastahyeti**
This PR makes a few changes. Let me know if you'd like me to split this up into multiple, smallers PRs.

- Adds a `.gitignore` file to avoid checking in unnecessary changes to the Xcode project.
- Adds targets to the Xcode project for the examples (keyboard/mouse).
- Fixes up the keyboard example to cleanup the virtual keyboard when the program exits.
- Extracts constants used in kernel *and* userland into `foohid_user_kernel_shared.h`
- Adds externally callable methods for opening/closing the user-client more cleanly.
- Adds some more cleanup code to the user client so it gets properly removed from registry when finished.

I also cleaned up a bunch of garbage whitespace throughout the project. This makes the diff kind of hard to read on GitHub. Appending [`?w=1`](https://github.com/unbit/foohid/pull/8/files?w=1) to the URL hides the whitespace changes from the diff, making it easier to read.

I'll be opening a separate PR to implement the `setReport` stuff.

### CODE REVIEW

**Comment by mastahyeti** [comment]
[Here's](https://github.com/mastahyeti/foohid/compare/master...mastahyeti:set-report) a branch adding `setReport` support. It's based on this branch, so I can't cleanly open a PR until this one merges.

**Comment by aldur** [comment]
Thanks! I reviewed the changed files, but I didn't test the new `kext` (I don't have a macOS install available ATM). @unbit, what do you think?

File: foohid/foohid_userclient.cpp [hunk 1]
```diff
@@ -100,94 +143,127 @@ IOReturn it_unbit_foohid_userclient::sMethodList(it_unbit_foohid_userclient *tar
     return target->methodList(arguments);
 }
 
+IOReturn it_unbit_foohid_userclient::methodOpen() {
+    if (m_hid_provider == NULL || isInactive()) {
+        LogD("methodOpen->kIOReturnNotAttached");
+        return kIOReturnNotAttached;
+    }
+
+    if (!m_hid_provider->open(this)) {
```
Comment (mastahyeti) [inline foohid/foohid_userclient.cpp]: We probably don't want to call `open()` on the provider. The default IOService implementation only allows open open client at a time. I don't think we actually need `methodOpen` and `methodClose`. Instead, we should just do some of the cleanup from `methodClose` in `clientClose` or `didTerminate`. I can make those changes to this PR tomorrow.

**Comment by unbit** [comment]
The patch seems good but i need to test it :) I will check it during the week-end

**Comment by mastahyeti** [comment]
@unbit Cool, thanks. Would you like me to include the `setReport` changes in this PR or do that in a separate PR once this one merges?

**Comment by mastahyeti** [comment]
This PR ended up mostly being aesthetic changes that aren't really important. I opened another PR (https://github.com/unbit/foohid/pull/9) to implement the `setReport` call, which is what I'm really interested in :smile:. I'm closing this PR for now.

**Pull request closed without merging**